src/components/AutoTagBuilder/AutoTagBuilder.int.part.tsx

```diff
@@ -0,0 +1,268 @@
+import * as React from 'react';
+import styled, { themed, css } from '../../utils/styled';
+import { FormContextProps } from '../../hoc/withFormContext';
+import { InteractiveListWrapperProps, InteractiveListDirection } from '../InteractiveList';
+import { InputChangeEvent } from '../../common';
+import { AutoTagBuilderProps, AutoTagBuilderState } from './AutoTagBuilder.types.part';
+import { debounce } from '../../utils';
+import { TagBuilder } from '../TagBuilder';
+import { Autocomplete, AutocompleteInputProps, AutosuggestSelectEvent } from '../Autocomplete';
+
+interface StyledAutoTagBuilderWrapperProps {
+  direction: InteractiveListDirection;
+}
+
+const StyledAutoTagBuilderWrapper = styled.ul<StyledAutoTagBuilderWrapperProps>(
+  themed(
+    ({ direction, theme: { ui1, ui4 } }) => css`
+      list-style: none;
+      width: 100%;
+      box-sizing: border-box;
+      box-shadow: none;
+      margin: 0;
+      padding: 0;
+      background: ${ui1};
+      border: 1px solid ${ui4};
+      ${direction === InteractiveListDirection.normal
+        ? 'border-top-color: transparent'
+        : 'border-bottom-color: transparent'};
+      max-height: 50vh;
+      position: absolute;
+      top: ${direction === InteractiveListDirection.normal ? '0' : '-100%'};
+      transform: translateY(${direction === InteractiveListDirection.normal ? 0 : -100}%);
+      overflow-y: auto;
+      z-index: 100;
+    `,
+  ),
+);
+
+// tslint:disable-next-line
+const NotOpenComponent = null;
+
+const InteractiveListWrapper: React.SFC<InteractiveListWrapperProps> = ({ border: _0, open, ...props }) =>
+  open ? <StyledAutoTagBuilderWrapper {...props} /> : NotOpenComponent;
+InteractiveListWrapper.displayName = 'InteractiveListWrapper';
+
+export class AutoTagBuilderInt<T> extends React.Component<
+  AutoTagBuilderProps<T> & FormContextProps,
+  AutoTagBuilderState<T>
+> {
+  private _fireOnInputChange: (q: string) => void;
+  private _enableInputChange: boolean;
+  private _inputNode: HTMLElement | null;
+
+  public constructor(props: AutoTagBuilderProps<T>) {
+    super(props);
+
+    const { value: nullableValue, defaultValue, onInputChange, delay = 0 } = this.props;
+    const value = nullableValue || defaultValue || [];
+    const valueMap = this.mapItemArray(value);
+
+    this.state = {
+      value: valueMap,
+      tagValue: value.map(x => this.getSuggestionValue(x)),
+      inputValue: '',
+      open: false,
```

Comment:
please name variables so that it's obvious which type they are of: opened or isOpen